compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -1162,6 +1176,76 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
         self.point_at_returns_when_relevant(&mut err, &obligation);
         err.emit();
     }
+
+    fn report_const_param_not_wf(
+        &self,
+        ty: Ty<'tcx>,
+        span: Span,
+    ) -> DiagnosticBuilder<'tcx, ErrorGuaranteed> {
+        let diag = match ty.kind() {
+            ty::Param(_) => {
+                // Const parameters may not have type parameters as their types,
+                // because we cannot be sure that the type parameter derives `PartialEq`
+                // and `Eq` (just implementing them is not enough for `structural_match`).
+                let mut diag = struct_span_err!(
+                    self.tcx.sess,
+                    span,
+                    E0741,
+                    "`{ty}` is not guaranteed to `#[derive(PartialEq, Eq)]`, so may not be \
+                    used as the type of a const generic parameter",
+                );
+                diag.note(
+                    "it is not currently possible to use a type parameter as the type of a \
+                    const parameter",
+                );
+                diag
+            }
+            ty::Float(_) => {
+                struct_span_err!(
+                    self.tcx.sess,
+                    span,
+                    E0741,
+                    "`{ty}` is forbidden as the type of a const generic parameter",
+                )
+            }
+            ty::FnPtr(_) => {
+                struct_span_err!(
+                    self.tcx.sess,
+                    span,
+                    E0741,
+                    "using function pointers as const generic parameters is forbidden",
+                )
+            }
+            ty::RawPtr(_) => {
+                struct_span_err!(
+                    self.tcx.sess,
+                    span,
+                    E0741,
+                    "using raw pointers as const generic parameters is forbidden",
+                )
+            }
+            ty::Adt(_, _) => {
+                // We should probably see if we're *allowed* to derive `ConstParamTy` on the type...
+                struct_span_err!(
+                    self.tcx.sess,
+                    span,
+                    E0741,
+                    "`{ty}` must be annotated with `#[derive(ConstParamTy)]` to be used as \
```

Comment:
This isnt really correct since you can also implement it manually, I think it would make sense to just say "`{ty}` must implement `ConstParamTy` to be used as the type of a const generic parameter" with a help to do `+ #[derive(ConstParamTy)]`

Reply:
ah yes, you're right -- forgot that the plan was to make it manually implementable.